src/workerd/jsg/async-context.h

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2017-2022 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+#pragma once
+
+#include "jsg.h"
+#include <v8.h>
+
+namespace workerd::jsg {
+
+class AsyncResource {
+  // Provides for basic internal async context tracking. Eventually, it is expected that
+  // this will be provided by V8 assuming that the AsyncContext proposal advances through
+  // TC-39. For now, however, we implement a model that is very similar to that implemented
+  // by Node.js.
+  //
+  // The term "resource" here comes from Node.js, which really doesn't take the time to
```

Comment:
I'm still somewhat unclear on what constitutes a "resource". You give the examples of promises and timers (by which I assume you mean `setTimeout()`). In both of these cases, the "resource" is created at the same time as the callback is specified (i.e. when `then()` or `setTimeout()` is called). Therefore, the callback inherits the async context of the scope which registered the callback. That makes perfect sense to me.

However, looking at your chances in `HTMLRewriter`, I see you create a single resource for the rewriter in its cosntructor, and all callbacks are called under that context. This does not make sense to me, because it means the async context is not inherited lexically.

How does this work in Node? If I create a stream in one async context, and then I add an event listener callback on the stream from some other async context, which context is the callback called in? IMO it should be the context from which I added the event listener. It doesn't make any sense to me that the context would be associated with the stream object itself.

Reply:
Node.js supports multiple models, providing `AsyncResource` to allow users to choose.

For instance, many thing (including a Node.js stream) is essentially just an EventEmitter instance. EventEmitters are generally purely synchronous objects.

```
const ee = new EventEmitter();
ee.on('foo', () => { 
  // Executes in the same execution context as the call to ee.emit('foo');
});
ee.emit('foo');
```

So, for instance, suppose I add an event listener within a promise continuation, but emit the event at some other point... the execution context for the handler is still the same context as the call to ee.emit('foo');

```
const ee = new EventEmitter();

somePromise.then(() => {
  ee.addEventListener('foo', () => {
    // Executes in the same execution context as the call to ee.emit('foo');
  });
});

setTimeout(() => {
  ee.emit('foo');
}, 10);
```

What if you want the `foo` event handler to run within the execution context of the *promise continuation* in which is was added (which is what I think most would actually want). To accomplish that, we use an `AsyncResource`

```
const ee = new EventEmitter();

somePromise.then(() => {
  ee.addEventListener('foo', AsyncResource::bind('foo', () => {
    // Executes in the AsyncResource's execution context, which is subordinate to the
    // promise continuations context 
  }));
});

setTimeout(() => {
  ee.emit('foo');
}, 10);
```

`AsyncResource` provides additional flexibility, however, in that users to create their own subclasses to have even greater control.

```js
const Foo extends AsyncResource {
  constructor(callback1, callback2) {
    super('Foo');
    this.#callback1 = callback1;
    this.#callback2 = callback2;
  }

  doSomething() {
    this.runInAsyncContext(() => {
      this.#callback1(); // Always invoked in this objects scope
    });
    this.#callback2(); // Always executes in the callers scope
  }
}
```

Here's a more complete example that shows use of `AsyncResource` for a pool implementation: https://nodejs.org/dist/latest-v19.x/docs/api/async_context.html#using-asyncresource-for-a-worker-thread-pool

Reply:
With regards to `HtmlRewriter` specifically, we absolutely can tweak how the async context is captured here. The approach I take in the PR is only one approach. 

With the way things are currently implemented, the default async context for an `HtmlRewriter` instance is whatever context it was created in. `AsyncResource` can be used with a specific `on(...)` call. For example:

```
class ElementHandler extends AsyncResource {
  element(element) {
    return this.runInAsyncContext(() => {
       // handle the element, return the result
       // runs in the context in which ElementHandler was created.
    });
  }
}

const rewriter = new HtmlRewriter();

als.run(123, () => { rewriter.on('*', new ElementHandler()); });

event.respondWith(rewriter.transform(res));
```

This would be the approach that Node.js would take in general. But, since `HtmlRewriter` is our own thing we definitely have flexibility here. We can choose to capture the context when `on(...)` is called rather than when the `HtmlRewriter` is created if that's your preference.

Reply:
Btw, fwiw, the reason I went with the approach I did (capturing the async resource when `HtmlRewriter` is created) is that it was way easier. Capturing it automatically when the various handler callbacks are registered is a rather more complicated change without a larger refactor of the code. If you prefer changing this then I'll drop the HtmlRewriter changes from this PR and will handle those separately.

Reply:
Marking resolved since the HtmlRewriter changes were removed here. We'll pick up consideration of what model that should use in a separate discussion